etna/transforms/math/lags.py

```diff
@@ -96,3 +98,148 @@ def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
     def get_regressors_info(self) -> List[str]:
         """Return the list with regressors created by the transform."""
         return [self._get_column_name(lag) for lag in self.lags]
+
+
+class ExogShiftTransform(IrreversibleTransform, FutureMixin):
+    """Shifts exogenous variables from a given dataframe."""
+
+    def __init__(self, lag: Union[int, Literal["auto"]], horizon: Optional[int] = None):
+        """Create instance of ExogShiftTransform.
+
+        Parameters
+        ----------
+        lag:
+            value for shift estimation
+
+            * if set to `int` all exogenous variables will be shifted `lag` steps forward;
+
+            * if set to `auto` minimal shift will be estimated for each variable based on
+              the prediction horizon and available timeline
+
+        horizon:
+            prediction horizon. Mandatory when set to `lag="auto"`, ignored otherwise
+        """
+        super().__init__(required_features="all")
+
+        self.lag: Optional[int] = None
+        self.horizon: Optional[int] = None
+        self.auto = False
+
+        self._created_regressors: Optional[List[str]] = None
+        self._exog_shifts: Optional[Dict[str, int]] = None
+
+        if isinstance(lag, int):
+            if lag <= 0:
+                raise ValueError(f"{type(self).__name__} works only with positive lags values, {lag} given")
+            self.lag = lag
+
+        else:
+            if horizon is None:
+                raise ValueError("`horizon` should be specified when using `auto`!")
+
+            if horizon < 1:
+                raise ValueError(f"{type(self).__name__} works only with positive horizon values, {horizon} given")
+
+            self.horizon = horizon
+            self.auto = True
+
+    def _fit(self, df: pd.DataFrame) -> "ExogShiftTransform":
+        """Estimate shifts for exogenous variables.
+
+        Parameters
+        ----------
+        df:
+            dataframe with data.
+
+        Returns
+        -------
+        result: ExogShiftTransform
+        """
+        feature_names = self._get_feature_names(df=df)
+
+        self._exog_shifts = dict()
+        self._created_regressors = []
+
+        for feature_name in feature_names:
+            shift = self._estimate_shift(df=df, feature_name=feature_name)
+            self._exog_shifts[feature_name] = shift
+
+            if shift > 0:
+                self._created_regressors.append(f"{feature_name}_shift_{shift}")
+
+        return self
+
+    @staticmethod
+    def _get_feature_names(df: pd.DataFrame) -> List[str]:
+        """Return the names of exogenous variables."""
+        names = set(df.columns.get_level_values("feature"))
```

Comment:
What about target components and quantiles?

Reply:
As I understand, `match_target_components` shouldn't be used now. The info about target components is stored in `TSDataset`, it isn't required to match any string pattern.